src/Riok.Mapperly/Descriptors/MappingBodyBuilder/NewInstanceObjectPropertyMappingBodyBuilder.cs

```diff
@@ -197,40 +197,72 @@ private static void BuildInitOnlyPropertyMappings(NewInstanceMappingBuilderConte
-                ctx.IgnoredSourcePropertyNames,
-                StringComparer.OrdinalIgnoreCase,
-                out var sourcePath))
-            {
-                if (!parameter.IsOptional)
-                    return false;
-
-                skippedOptionalParam = true;
-                continue;
+        {
+            if (ctx.PropertyConfigsByRootTargetName.Remove(parameter.Name, out var propertyConfigs))
+            {
+                foreach (var config in propertyConfigs.OrderBy(x => x.Target.Count))
+                {
+                    if (!PropertyPath.TryFind(ctx.Mapping.SourceType, config.Source, out var sourcePath))
+                    {
+                        ctx.BuilderContext.ReportDiagnostic(
+                            DiagnosticDescriptors.ConfiguredMappingSourcePropertyNotFound,
+                            string.Join(PropertyPath.PropertyAccessSeparator, config.Source),
+                            ctx.Mapping.SourceType);
+                        continue;
+                    }
+
+                    TryAddMapping(ctx, parameter, sourcePath, ref skippedOptionalParam, ref constructorParameterMappings, ref mappedTargetPropertyNames);
+                }
             }
-
-            // nullability is handled inside the property mapping
-            var paramType = parameter.Type.WithNullableAnnotation(parameter.NullableAnnotation);
-            var delegateMapping = ctx.BuilderContext.FindMapping(sourcePath.MemberType, paramType)
-                ?? ctx.BuilderContext.FindOrBuildMapping(sourcePath.Member.Type.NonNullable(), paramType);
-            if (delegateMapping == null)
-            {
-                if (!parameter.IsOptional)
-                    return false;
-
-                skippedOptionalParam = true;
-                continue;
+            else
+            {
+                if (!PropertyPath.TryFind(
+                    ctx.Mapping.SourceType,
+                    MemberPathCandidateBuilder.BuildMemberPathCandidates(parameter.Name),
+                    ctx.IgnoredSourcePropertyNames,
+                    StringComparer.OrdinalIgnoreCase,
+                    out var sourcePath))
+                {
+                    if (!parameter.IsOptional)
+                        return false;
+
+                    skippedOptionalParam = true;
+                    continue;
+                }
+
+                TryAddMapping(ctx, parameter, sourcePath, ref skippedOptionalParam, ref constructorParameterMappings, ref mappedTargetPropertyNames);
             }
-
-            var propertyMapping = new NullPropertyMapping(delegateMapping, sourcePath, ctx.BuilderContext.GetNullFallbackValue(paramType));
-            var ctorMapping = new ConstructorParameterMapping(parameter, propertyMapping, skippedOptionalParam);
-            constructorParameterMappings.Add(ctorMapping);
-            mappedTargetPropertyNames.Add(parameter.Name);
         }
 
         return true;
+
+        static bool TryAddMapping(

```

Comment:
IMO instead of the ref parameterised method you proposed the following approach could simplify the code and improve the overall readability:
Add a method with the following signature:
```csharp
private static bool TryFindConstructorParameterSourcePath(
        NewInstanceMappingBuilderContext ctx,
        IParameterSymbol parameter,
        [NotNullWhen(true)] out PropertyPath? sourcePath)
```
which tries to get the `MapPropertyAttribute` from `PropertyConfigsByRootTargetName`, validates it, creates diagnostics as needed (see `BuildInitPropertyMapping `) and resolves the source property path or if `PropertyConfigsByRootTargetName` doesn't return an entry, try to resolve the source property directly (as currently implemented in the else branch).
In the `foreach (var parameter in ctor.Parameters)` loop then call this new method. If true is returned execute the code you currently have in `TryAddMapping`. If false is returned, check if the parameter is optional, if it is, set `skippedOptionalParam` to true and continue, if it isn't optional return false.